internal/local/archive.go

```diff
@@ -0,0 +1,207 @@
+package local
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// set of supported archive file extensions
+const (
+	extTar   = ".tar"
+	extTarGz = ".tar.gz"
+	extTgz   = ".tgz"
+	extZip   = ".zip"
+)
+
+var (
+	supportedExts = []string{extZip, extTar, extTgz, extTarGz}
+)
+
+// ArchiveReader provides sequential access to the contents of the supported archive formats
+type ArchiveReader interface {
+	Next() (FileHeader, error)
+	Read(b []byte) (int, error)
+}
+
+// FileHeader holds the path and file information
+type FileHeader struct {
+	Path string
+	Info os.FileInfo
+}
+
+func newArchiveReader(path string, file *os.File) (ArchiveReader, error) {
+	ext := strings.ToLower(filepath.Ext(path))
+
+	if ext == extZip {
+		return newZipReader(file)
+	}
+
+	if ext == extTar {
+		return newTarReader(file), nil
+	}
+
+	if ext == extTgz || strings.HasSuffix(path, extTarGz) {
+		return newGzipReader(file)
+	}
+
+	if ext == "" {
+		return newDirReader(path)
+	}
+
+	return nil, errUnknownArchiveExtension(path)
+}
+
+func errUnknownArchiveExtension(path string) error {
+	return fmt.Errorf(
+		"failed to read archive file at %s: unsupported format, use one of [%s] instead",
+		path,
+		strings.Join(supportedExts, ", "),
+	)
+}
```

Comment:
[nit] this seems unnecessary. Is the sole purpose of this func for tests?

Reply:
the sole purpose of this func was for reducing the visual load when looking at the calling code... i can move it back, these things just tend to help me